### ISSUE TITLE
Tests Take an Impractically Long Time To Run

### ISSUE THREAD

**Issue by Miserlou** [AlexsLemonade/refinebio#106]
<img width="1296" alt="screen shot 2018-01-29 at 2 28 58 pm" src="https://user-images.githubusercontent.com/139987/35530019-d00fb792-0500-11e8-9e79-09f3c8b886ac.png">

The build time is now `1:31:04`

This is almost entirely because of `data_refinery_workers.processors.test_salmon.SalmonTestCase.test_success`.

One possible solution might be to [only run the longer tests on the dev branch](https://discuss.circleci.com/t/multiple-builds-for-one-repo/662/3) rather than feature branches so that feature builds happen faster, but the full suite is run when the test is run as a dev PR.

Another solution is to find a much faster Salmon test success, or to figure out how to run Salmon as a dry run.

**Comment by Miserlou** [comment]
I've added a small convenience hack to the CircleCI script: If you include the word `noslow` in your commit message, it will skip the `test_success` case which takes so darn long.

**Comment by Miserlou** [comment]
Using `noslow` cuts about an hour off the build time, but it's still about 20-25 minutes longer than it needs to be, where most of the time is eaten up by Circle having to download the `ccdl/data_refinery_worker_base` image.

Since we need to document and refactor this image anyway - as mentioned in #109 - is there any chance we can (significantly) slim this image down? At least for a test or noslow image? I'd really like to get the build time down to a few minutes other wise it's just not very practical.

**Comment by Miserlou** [comment]
One of these things is not like the others..

```bash
busybox                          latest              f9b6f7f7b9d3        2 weeks ago         1.14MB
centos                           centos7             ff426288ea90        3 weeks ago         207MB
rabbitmq                         3                   d53246bdedd1        5 weeks ago         127MB
postgres                         9.6.6               5579c7505b1b        7 weeks ago         268MB
debian                           8                   2fe79f06fa6d        7 weeks ago         123MB
ccdl/data_refinery_worker_base   latest              ed2794a5b1ff        2 months ago        22.5GB
djenriquez/nomad                 v0.6.0              a88ea20dbdd3        6 months ago        59.2MB
python                           3.6.1-slim          93b738c915e5        7 months ago        200MB
vancluever/nomad                 latest              226ed5ddb75d        8 months ago        45.3MB
```

From `docker history ccdl/data_refinery_worker_base`, it looks like most of the heft comes from this step:

`
<missing>           6 months ago        /bin/sh -c Rscript r_dependencies.R             20.8GB
`

Do we really need _all_ of these dependencies?

```R
# Turn warnings into errors because biocLite throws warnings instead
# of error if it fails to install something.
options(warn=2)

install.packages('ff', repos='https://cloud.r-project.org/')
install.packages('XML', repos='https://cloud.r-project.org/')
install.packages('RCurl', repos='https://cloud.r-project.org/')
install.packages('RSQLite', repos='https://cloud.r-project.org/')
install.packages('tibble', repos='https://cloud.r-project.org/')
install.packages('xtable', repos='https://cloud.r-project.org/')
install.packages('pkgconfig', repos='https://cloud.r-project.org/')

source('https://bioconductor.org/biocLite.R')
bioconductor.packages <- c(
  'oligo',
  'Biobase',
  'SCAN.UPC',
  'affy',
  'affyio',
  'AnnotationDbi',
  'zlibbioc',
  'preprocessCore',
  'genefilter',
  'sva',
  'org.Hs.eg.db',
  'org.Mm.eg.db',
  'org.Dm.eg.db',
  'org.Ce.eg.db',
  'org.Bt.eg.db',
  'org.Cf.eg.db',
  'org.Gg.eg.db',
  'org.Rn.eg.db',
  'org.Ss.eg.db',
  'org.Dr.eg.db'
)

lapply(bioconductor.packages, biocLite)


packages <- row.names(available.packages(contriburl="http://brainarray.mbni.med.umich.edu/bioc/src/contrib/"))

## This narrows down from all the files gotten just having entrezg in the name
entrez.packages <- packages[grep("entrezg", packages)]

install.packages(entrez.packages, repos="http://brainarray.mbni.med.umich.edu/bioc/", type="source")

platform.designs <- c(
  "pd.081229.hg18.promoter.medip.hx1",
  "pd.2006.07.18.hg18.refseq.promoter",
  "pd.2006.07.18.mm8.refseq.promoter",
  "pd.2006.10.31.rn34.refseq.promoter",
  "pd.ag",
  "pd.aragene.1.0.st",
  "pd.aragene.1.1.st",
  "pd.atdschip.tiling",
  "pd.ath1.121501",
  "pd.barley1",
  "pd.bovgene.1.0.st",
  "pd.bovgene.1.1.st",
  "pd.bovine",
  "pd.bsubtilis",
  "pd.cangene.1.0.st",
  "pd.cangene.1.1.st",
  "pd.canine",
  "pd.canine.2",
  "pd.celegans",
  "pd.charm.hg18.example",
  "pd.chicken",
  "pd.chigene.1.0.st",
  "pd.chigene.1.1.st",
  "pd.chogene.2.0.st",
  "pd.chogene.2.1.st",
  "pd.citrus",
  "pd.clariom.d.human",
  "pd.clariom.s.human",
  "pd.clariom.s.human.ht",
  "pd.clariom.s.mouse",
  "pd.clariom.s.mouse.ht",
  "pd.clariom.s.rat",
  "pd.clariom.s.rat.ht",
  "pd.cotton",
  "pd.cyngene.1.0.st",
  "pd.cyngene.1.1.st",
  "pd.cyrgene.1.0.st",
  "pd.cyrgene.1.1.st",
  "pd.cytogenetics.array",
  "pd.drogene.1.0.st",
  "pd.drogene.1.1.st",
  "pd.drosgenome1",
  "pd.drosophila.2",
  "pd.ecoli",
  "pd.e.coli.2",
  "pd.ecoli.asv2",
  "pd.elegene.1.0.st",
  "pd.elegene.1.1.st",
  "pd.equgene.1.0.st",
  "pd.equgene.1.1.st",
  "pd.feinberg.hg18.me.hx1",
  "pd.feinberg.mm8.me.hx1",
  "pd.felgene.1.0.st",
  "pd.felgene.1.1.st",
  "pd.fingene.1.0.st",
  "pd.fingene.1.1.st",
  "pd.genomewidesnp.5",
  "pd.genomewidesnp.6",
  "pd.guigene.1.0.st",
  "pd.guigene.1.1.st",
  "pd.hc.g110",
  "pd.hg18.60mer.expr",
  "pd.hg.focus",
  "pd.hg.u133a",
  "pd.hg.u133a.2",
  "pd.hg.u133a.tag",
  "pd.hg.u133b",
  "pd.hg.u133.plus.2",
  "pd.hg.u219",
  "pd.hg.u95a",
  "pd.hg.u95av2",
  "pd.hg.u95b",
  "pd.hg.u95c",
  "pd.hg.u95d",
  "pd.hg.u95e",
  "pd.hta.2.0",
  "pd.ht.hg.u133a",
  "pd.ht.hg.u133.plus.pm",
  "pd.ht.mg.430a",
  "pd.hu6800",
  "pd.huex.1.0.st.v2",
  "pd.hugene.1.0.st.v1",
  "pd.hugene.1.1.st.v1",
  "pd.hugene.2.0.st",
  "pd.hugene.2.1.st",
  "pd.maize",
  "pd.mapping250k.nsp",
  "pd.mapping250k.sty",
  "pd.mapping50k.hind240",
  "pd.mapping50k.xba240",
  "pd.margene.1.0.st",
  "pd.margene.1.1.st",
  "pd.medgene.1.0.st",
  "pd.medgene.1.1.st",
  "pd.medicago",
  "pd.mg.u74a",
  "pd.mg.u74av2",
  "pd.mg.u74b",
  "pd.mg.u74bv2",
  "pd.mg.u74c",
  "pd.mg.u74cv2",
  "pd.mirna.1.0",
  "pd.mirna.2.0",
  "pd.mirna.3.0",
  "pd.mirna.3.1",
  "pd.mirna.4.0",
  "pd.moe430a",
  "pd.moe430b",
  "pd.moex.1.0.st.v1",
  "pd.mogene.1.0.st.v1",
  "pd.mogene.1.1.st.v1",
  "pd.mogene.2.0.st",
  "pd.mogene.2.1.st",
  "pd.mouse430.2",
  "pd.mouse430a.2",
  "pd.mta.1.0",
  "pd.mu11ksuba",
  "pd.mu11ksubb",
  "pd.nugo.hs1a520180",
  "pd.nugo.mm1a520177",
  "pd.ovigene.1.0.st",
  "pd.ovigene.1.1.st",
  "pd.pae.g1a",
  "pd.plasmodium.anopheles",
  "pd.poplar",
  "pd.porcine",
  "pd.porgene.1.0.st",
  "pd.porgene.1.1.st",
  "pd.rabgene.1.0.st",
  "pd.rabgene.1.1.st",
  "pd.rae230a",
  "pd.rae230b",
  "pd.raex.1.0.st.v1",
  "pd.ragene.1.0.st.v1",
  "pd.ragene.1.1.st.v1",
  "pd.ragene.2.0.st",
  "pd.ragene.2.1.st",
  "pd.rat230.2",
  "pd.rcngene.1.0.st",
  "pd.rcngene.1.1.st",
  "pd.rg.u34a",
  "pd.rg.u34b",
  "pd.rg.u34c",
  "pd.rhegene.1.0.st",
  "pd.rhegene.1.1.st",
  "pd.rhesus",
  "pd.rice",
  "pd.rjpgene.1.0.st",
  "pd.rjpgene.1.1.st",
  "pd.rn.u34",
  "pd.rta.1.0",
  "pd.rusgene.1.0.st",
  "pd.rusgene.1.1.st",
  "pd.s.aureus",
  "pd.soybean",
  "pd.soygene.1.0.st",
  "pd.soygene.1.1.st",
  "pd.sugar.cane",
  "pd.tomato",
  "pd.u133.x3p",
  "pd.vitis.vinifera",
  "pd.wheat",
  "pd.xenopus.laevis",
  "pd.x.laevis.2",
  "pd.x.tropicalis",
  "pd.yeast.2",
  "pd.yg.s98",
  "pd.zebgene.1.0.st",
  "pd.zebgene.1.1.st",
  "pd.zebrafish"
)

lapply(platform.designs, biocLite)
```

Of these, what's the minimum required to run our tests?

**Comment by Miserlou** [comment]
Since apparently it is a goal to split out the worker image into separate images based on functionality, I think we can now safely split this into four images:

- Base
   - Salmon
   - Scan-Test
        - Scan-Full

Where Scan-Test has only the platform design needed for a test image, whereas Scan-Full will have all of the platforms listed above.

Does this make sense? @kurtwheeler

**Comment by cgreene** [comment]
I'm not sure about the status of this. It does seem like we need tests working in a practical way for the Keytar Kurt milestone. Is this something that can be closed, or is this still an issue?

**Comment by cgreene** [comment]
My impression is that #262 addresses this issue. Putting back into Keytar since it appears to be largely resolved by that existing PR and #256 which is also tagged for Keytar.
